app/auth/auth.go

```diff
@@ -0,0 +1,98 @@
+package auth
+
+import (
+	"context"
+	"net/http"
+
+	"github.com/lbryio/lbrytv/app/sdkrouter"
+	"github.com/lbryio/lbrytv/app/wallet"
+	"github.com/lbryio/lbrytv/internal/ip"
+	"github.com/lbryio/lbrytv/internal/monitor"
+	"github.com/lbryio/lbrytv/models"
+
+	"github.com/gorilla/mux"
+)
+
+var logger = monitor.NewModuleLogger("auth")
+
+const ContextKey = "user"
+
+func FromRequest(r *http.Request) Result {
+	v := r.Context().Value(ContextKey)
+	if v == nil {
+		panic("Auth middleware was not applied")
+	}
+	return v.(Result)
+}
+
+// Provider tries to authenticate using the provided auth token
+type Provider func(token, metaRemoteIP string) Result
+
+// WalletAndInternalAPIProvider auths a user by hitting internal-api with the auth token
+// and matching the response to a local user. If auth is successful, the user will have a
+// lbrynet server assigned and a wallet that's created and ready to use.
+func NewWalletAndInternalAPIProvider(rt *sdkrouter.Router, internalAPIHost string) Provider {
+	return func(token, metaRemoteIP string) Result {
+		user, err := wallet.GetUserWithWallet(rt, internalAPIHost, token, metaRemoteIP)
+		res := NewResult(user, err)
+		if err == nil && user != nil && !user.LbrynetServerID.IsZero() && user.R != nil && user.R.LbrynetServer != nil {
+			res.SDKAddress = user.R.LbrynetServer.Address
+		}
+		return res
+	}
+}
+
+func Middleware(provider Provider) mux.MiddlewareFunc {
+	return func(next http.Handler) http.Handler {
+		return http.HandlerFunc(func(w http.ResponseWriter, r *http.Request) {
+			var res Result
+			if token, ok := r.Header[wallet.TokenHeader]; ok {
+				addr := ip.AddressForRequest(r)
+				res = provider(token[0], addr)
+				if res.err != nil {
+					logger.LogF(monitor.F{"ip": addr}).Debugf("error authenticating user")
+				}
+				res.authAttempted = true
+			}
+			next.ServeHTTP(w, r.Clone(context.WithValue(r.Context(), ContextKey, res)))
+		})
+	}
+}
+
+// wish i could make this non-exported, but then you can't create new providers outside the package
+// don't make this struct directly. instead use NewResult
+type Result struct {
```

Comment:
its an auth.Result - the result of authentication

i'm not sure the best way to do this. what i'd like is for it to either return a user, or an error, or neither. if neither, no authentication happened (aka there was no auth token). if error, then it happened and errors. if user, then it happened and was successful.

i'm open to ideas for how to do this better

Reply:
I would just return `nil, nil` for when there is neither an error nor a user

Reply:
but it also returns extra stuff, like sdk address. and might return more in the future. and it makes for a more complex interface when you're returning a bunch of values instead of just one